File: python/xfa/02_data_binding/02_data_binding.py
# ============================================================================
#  02_data_binding -- Python (ctypes) port of
#  examples\delphi\xfa\02_data_binding\02_data_binding.dpr
#
#  LumasPDF XFA "flavor tour" example 2 of 10 -- DATA BINDING. Demonstrates
#  the three binding modes an XFA form mixes in practice, all against one
#  realistic, genuinely nested <xfa:datasets> packet:
#    1. Implicit binding (by-name, no <bind> element, containment-aware).
#    2. Explicit <bind match="dataRef" ref="$data...."/> against a nested
#       SOM path (3 and 4 levels deep).
#    3. <bind match="none"/> -- pure literal, unaffected by same-named data.
#
#    pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
#    pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
#
#  Packets are read pre-split from their own .template.xml/.datasets.xml
#  files (extracted from the combined .xdp by this session's one-off
#  split_xfa_packets tool) -- no XML library needed in this driver.
# ============================================================================
import os
import sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

HERE = os.path.dirname(os.path.abspath(__file__))


def render(template_path, datasets_path, out_pdf_path):
    template = open(template_path, "rb").read()
    datasets = open(datasets_path, "rb").read()
    print("template packet bytes:", len(template))
    print("datasets packet bytes:", len(datasets))

    pdf = L.pdfNewPDF()
    if not pdf:
        print("pdfNewPDF FAILED")
        return -100
    try:
        if L.pdfCreateNewPDFA(pdf, out_pdf_path.encode("latin-1")) == 0:
            print("pdfCreateNewPDFA FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"template", template, len(template))
        print("pdfCreateXFAStreamA(template) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(template) FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"datasets", datasets, len(datasets))
        print("pdfCreateXFAStreamA(datasets) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(datasets) FAILED")
            return -100

        rc = L.pdfRenderXFAForm(pdf)
        print("pdfRenderXFAForm ->", rc)
        if rc < 1:
            print("pdfRenderXFAForm FAILED, code", rc)
            return rc

        if L.pdfCloseFile(pdf) == 0:
            print("pdfCloseFile FAILED")
            return -101

        print("OK: wrote", out_pdf_path)
        return rc
    finally:
        L.pdfDeletePDF(pdf)


def main():
    print("LumasPDF XFA flavor tour -- example 2/10: Data Binding")
    print("(implicit by-name + explicit dataRef SOM path + match=none literal)")
    print()

    template_path = os.path.join(HERE, "02_data_binding.template.xml")
    datasets_path = os.path.join(HERE, "02_data_binding.datasets.xml")
    out_pdf_path = os.path.join(HERE, "02_data_binding.render.pdf")

    print("=== 02_data_binding -> 02_data_binding.render.pdf ===")
    rc = render(template_path, datasets_path, out_pdf_path)
    print()
    print("RESULT|02_data_binding=", rc)

    print()
    print("Expected resolved values (checked against the rendered PDF page 1")
    print("content stream via pypdf -- adjacent Tj caption/value pairs):")
    print('  Customer Name (implicit)              : "Acme Robotics LLC"')
    print('  Account ID (implicit)                  : "ACCT-88213"')
    print('  Street (implicit, nested subform)      : "500 Innovation Way"')
    print('  State (implicit, nested subform)       : "IL"')
    print('  Zip (implicit, nested subform)         : "62704"')
    print('  Shipping City (explicit dataRef, 3 deep): "Springfield"')
    print('  Primary Contact Email (dataRef, 4 deep) : "ap@acmerobotics.example"')
    print('  Status (match="none", literal)          : "Active - Verified"')
    print('    (NOT "PENDING_CLOSURE", the same-named trap node in datasets)')


if __name__ == "__main__":
    main()

File: csharp/xfa/02_data_binding/02_data_binding.cs
// ============================================================================
//  02_data_binding -- C# port of examples\delphi\xfa\02_data_binding
//  XFA "flavor tour" example 2 of 10: the three data-binding modes an XFA
//  form mixes in practice, all against one realistic, genuinely nested
//  <xfa:datasets> packet -- implicit (by-name, no <bind>), explicit
//  (<bind match="dataRef" ref="$data...."/> against a deep SOM path), and
//  <bind match="none"/> (pure literal, unaffected by same-named data).
//
//  Same pipeline as every example in this tour:
//    pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
//    pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
//
//  Does not rebuild LumasPdf.dll -- links against the generated P/Invoke
//  wrapper wrappers\dotnet\LumasPdf.cs (LumasPdf.Net.dll).
// ============================================================================
using System;
using System.IO;
using System.Runtime.InteropServices;
using LumasPdfSdk;

class DataBinding
{
    static int RenderExample(string exeDir, string baseName, string outPdfPath)
    {
        string templatePath = Path.Combine(exeDir, baseName + ".template.xml");
        string datasetsPath = Path.Combine(exeDir, baseName + ".datasets.xml");

        Console.WriteLine("=== " + baseName + " -> " + Path.GetFileName(outPdfPath) + " ===");
        if (!File.Exists(templatePath)) { Console.WriteLine("FILE-NOT-FOUND: " + templatePath); return -100; }

        byte[] templateBuf = File.ReadAllBytes(templatePath);
        byte[] datasetsBuf = File.Exists(datasetsPath) ? File.ReadAllBytes(datasetsPath) : new byte[0];
        Console.WriteLine("template packet bytes: " + templateBuf.Length);
        Console.WriteLine("datasets packet bytes: " + datasetsBuf.Length);

        IntPtr pdf = LumasPdf.pdfNewPDF();
        if (pdf == IntPtr.Zero) { Console.WriteLine("pdfNewPDF FAILED"); return -100; }
        try
        {
            if (!LumasPdf.pdfCreateNewPDFW(pdf, outPdfPath)) { Console.WriteLine("pdfCreateNewPDFA FAILED"); return -100; }

            GCHandle hTpl = GCHandle.Alloc(templateBuf, GCHandleType.Pinned);
            int idx;
            try { idx = LumasPdf.pdfCreateXFAStreamW(pdf, "template", hTpl.AddrOfPinnedObject(), (uint)templateBuf.Length); }
            finally { hTpl.Free(); }
            Console.WriteLine("pdfCreateXFAStreamA(template) -> index " + idx);
            if (idx < 0) { Console.WriteLine("pdfCreateXFAStreamA(template) FAILED"); return -100; }

            if (datasetsBuf.Length > 0)
            {
                GCHandle hDs = GCHandle.Alloc(datasetsBuf, GCHandleType.Pinned);
                try { idx = LumasPdf.pdfCreateXFAStreamW(pdf, "datasets", hDs.AddrOfPinnedObject(), (uint)datasetsBuf.Length); }
                finally { hDs.Free(); }
                Console.WriteLine("pdfCreateXFAStreamA(datasets) -> index " + idx);
                if (idx < 0) { Console.WriteLine("pdfCreateXFAStreamA(datasets) FAILED"); return -100; }
            }
            else
            {
                Console.WriteLine("(no datasets packet found -- template-only render)");
            }

            int rc = LumasPdf.pdfRenderXFAForm(pdf);
            Console.WriteLine("pdfRenderXFAForm -> " + rc + " (expected: page count >= 1)");
            if (rc < 1) { Console.WriteLine("RENDER-FAILED, code " + rc); return rc; }

            if (!LumasPdf.pdfCloseFile(pdf)) { Console.WriteLine("pdfCloseFile FAILED"); return -101; }
            Console.WriteLine("OK: wrote " + outPdfPath + " (" + rc + " page(s))");
            return rc;
        }
        finally
        {
            LumasPdf.pdfDeletePDF(pdf);
        }
    }

    static void Main()
    {
        try
        {
            string exeDir = AppDomain.CurrentDomain.BaseDirectory;
            int rc = RenderExample(exeDir, "02_data_binding", Path.Combine(exeDir, "02_data_binding.render.pdf"));
            Console.WriteLine();
            Console.WriteLine("RESULT|02_data_binding=" + rc);
            if (rc >= 1)
            {
                Console.WriteLine("OK: rendered " + rc + " page(s). Open 02_data_binding.render.pdf and confirm:");
                Console.WriteLine("  Field                          Binding kind                  Resolved value");
                Console.WriteLine("  Customer Name                  implicit                      Acme Robotics LLC");
                Console.WriteLine("  Account ID                     implicit                      ACCT-88213");
                Console.WriteLine("  Street                         implicit (nested subform)      500 Innovation Way");
                Console.WriteLine("  State                          implicit (nested subform)      IL");
                Console.WriteLine("  Zip                            implicit (nested subform)      62704");
                Console.WriteLine("  Shipping City (SOM 3 deep)     explicit dataRef               Springfield");
                Console.WriteLine("  Primary Contact Email (4 deep) explicit dataRef               ap@acmerobotics.example");
                Console.WriteLine("  Status (literal, match=none)   match=\"none\"                   Active - Verified (NOT PENDING_CLOSURE)");
            }
            else
            {
                Console.WriteLine("FAILED, see errors above.");
            }
        }
        catch (Exception e)
        {
            Console.WriteLine("EXCEPTION: " + e.GetType().Name + ": " + e.Message);
        }
    }
}
